### DISCUSSION TITLE
Virtual (shared) GPU memory

### DISCUSSION THREAD

**Discussion by BaGRoS** [openai/whisper#259] [Ideas]
Please add the possibility to select ONLY GPU dedicated memory or dedicated and virtual memory (shared RAM).  My laptop only has 8GB of dedicated GPU memory and I cannot test the LARGE model. It is possible that using shared memory, the speed will be the same as CPU only or maybe something in the middle of CPU alone and GPU alone.

Many thanks,
BaGRoS

**Comment by ProjectEGU** [comment]
You can try [my fork](https://github.com/ProjectEGU/whisper-for-low-vram/tree/main/whisper) which implements the suggestion from [#180](https://github.com/openai/whisper/discussions/180#discussioncomment-3765617). It loads the encoder and decoder models separately, unloading the other when needed. With these changes, I can run the LARGE model on my 3060 Ti which has 8gb dedicated VRAM.

Install instructions:

```
pip uninstall whisper
pip install git+https://github.com/ProjectEGU/whisper-for-low-vram.git
```

**Reply by BaGRoS** [reply]
Later this can be used like a normal whisper? So for an external program it doesn't matter which version is used?

**Reply by ProjectEGU** [reply]
> Later this can be used like a normal whisper? So for an external program it doesn't matter which version is used?

The usage is exactly the same as normal whisper, so there's no need to make any changes to your software.

**Reply by retrotails** [reply]
I can confirm this works on my 8GB vega 56 which is great, although it's worth noting it uses a lot more system memory and sometimes the process is killed for using all 16GB. there doesn't seem to be much of a performance penalty, at least.

**Reply by FurkanGozukara** [reply]
> You can try [my fork](https://github.com/ProjectEGU/whisper-for-low-vram/tree/main/whisper) which implements the suggestion from [#180](https://github.com/openai/whisper/discussions/180#discussioncomment-3765617). It loads the encoder and decoder models separately, unloading the other when needed. With these changes, I can run the LARGE model on my 3060 Ti which has 8gb dedicated VRAM.
> 
> Install instructions:
> 
> ```
> pip uninstall whisper
> pip install git+https://github.com/ProjectEGU/whisper-for-low-vram.git
> ```

wow this looks awesome ty very much

**Reply by BaGRoS** [reply]
Are you going to update on a regular basis?

**Reply by amgmbs** [reply]
@jongwook are these changes implemented in "main" whisper today?

**Reply by ProjectEGU** [reply]
Hi softlypink,

My script is the same as the main whisper script, except for changes that make it runnable on low vram. It is periodically updated with changes from the main whisper repo. 

The problem you describe is likely an issue with the main whisper script. Please feel free to [start a new discussion](https://github.com/openai/whisper/discussions) about the issue or [raise a pull request](https://github.com/openai/whisper/pulls) with a fix.

Once the issue is fixed on the main script, it should be fixed in my forked repo as well.

**Reply by chislon** [reply]
> You can try [my fork](https://github.com/ProjectEGU/whisper-for-low-vram/tree/main/whisper) which implements the suggestion from [#180](https://github.com/openai/whisper/discussions/180#discussioncomment-3765617). It loads the encoder and decoder models separately, unloading the other when needed. With these changes, I can run the LARGE model on my 3060 Ti which has 8gb dedicated VRAM.
> 
> Install instructions:
> 
> ```
> pip uninstall whisper
> pip install git+https://github.com/ProjectEGU/whisper-for-low-vram.git
> ```

Results on a laptop with 4GB VRAM and integrated graphics using your modified model.py in-place of the one from 20230318, medium model VRAM usage reduced from 3.9GB to 3.3GB. 

So ProjectEGU's modifications actually allow for users with 4GB VRAM to run reliably with the medium model, which is quite nice. If someone didn't have integrated graphics they might not be able to run the medium model as-is without your changes.

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 531.41                 Driver Version: 531.41       CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                      TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3050 L...  WDDM | 00000000:01:00.0 Off |                  N/A |
| N/A   76C    P0               70W /  N/A|   3311MiB /  4096MiB |     90%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
```

**Reply by TrashcanCoder** [reply]
Works just file CUDA 11.7 (Windows) -  GTX3060Ti (8GB)
whisper-for-low-vram.git
